### PR TITLE
PWN-3560 - Сoin gecko as the second source of token prices

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3560

https://www.coingecko.com/en/api/documentation

## Description of Work

- feature toggle for Coin Gecko
- fetch prices for tokens with non-optimized ugly duck-taped way from CoinGecko (i haven't found any limitations for this API)

## Screenshots (Optional)
Screen Shot 2022-05-30 at 3 57 50 pm

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/user/repository/prices/sources/CoinGeckoApiClient.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            ?.inputStream()
+            ?.readBytes()
+            ?.let { jsonBytes ->
+                val listType = object : TypeToken<List<CoinGeckoTokenInformation>>() {}.type
+                gson.fromJson<List<CoinGeckoTokenInformation>>(String(jsonBytes), listType)
+            }
+            .orEmpty()
+    }
+
+    private fun mapSymbolToCoinGeckoTokenIds(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/repository/prices/sources/CoinGeckoApiClient.kt:79]: duck-taping:

- download a huge .json with all the token symbols and ids from /coins/list 
- save it locally
- map all the local user symbols (USDT, SOL, renBTC, ...) to ids from CoinGecko (usdt-stable, solana, bitcoin) to make the request

File: app/src/main/java/org/p2p/wallet/common/AppFeatureFlags.kt [hunk 2]
```diff
@@ -1,25 +1,14 @@
 package org.p2p.wallet.common
 
 import android.content.SharedPreferences
-import androidx.core.content.edit
 import org.p2p.wallet.BuildConfig
 
 private const val KEY_POLLING_ENABLED = "KEY_POLLING_ENABLED"
 private const val KEY_DEV_NET_ENABLED = "KEY_DEV_NET_ENABLED"
+private const val KEY_COIN_GECKO_ENABLED = "KEY_DEV_NET_ENABLED"
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/common/AppFeatureFlags.kt:8]: key copy pasted issue!

File: app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesRemoteRepository.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.p2p.wallet.common.AppFeatureFlags
+import org.p2p.wallet.home.model.TokenPrice
+import org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers
+import org.p2p.wallet.user.repository.prices.sources.CoinGeckoApiClient
+import org.p2p.wallet.user.repository.prices.sources.CryptoCompareApiClient
+import kotlinx.coroutines.withContext
+
+class TokenPricesRemoteRepository(
+    private val cryptoCompareApiClient: CryptoCompareApiClient,
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesRemoteRepository.kt:11]: i prefer to use same interface and provide it in TokenPricesRemoteRepository according to appFeatureFlags.useCoinGeckoForPrices

contact me for mode details if it is not clear

File: app/src/main/java/org/p2p/wallet/common/AppFeatureFlags.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/common/AppFeatureFlags.kt:8]: oh thanks man!!

File: app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesRemoteRepository.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesRemoteRepository.kt:11]: do you want to keep only TokenPricesRemoteRepository in this way or just inject one interface and choose repo implementation depending on feature flag?

Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesRemoteRepository.kt:11]: 2ed point
to pass interface and choose repo implementation depending on feature flag

File: app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesModule.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+object TokenPricesModule : InjectionModule {
+    override fun create() = module {
+        single {
+            val client = createOkHttpClient()
+                .addInterceptor(CompareTokenInterceptor())
+                .apply { if (BuildConfig.DEBUG) addInterceptor(httpLoggingInterceptor("CryptoCompare")) }
+                .build()
+
+            val compareApi = Retrofit.Builder()
```
Comment (kuznetsov) [inline app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesModule.kt:28]: `cryptoCompareApi` 🙏

File: app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesModule.kt [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                .baseUrl(get<Context>().getString(R.string.compareBaseUrl))
```
Comment (kuznetsov) [inline app/src/main/java/org/p2p/wallet/user/repository/prices/TokenPricesModule.kt:29]: `cryptoCompareBaseUrl`

**Review by kuznetsov — APPROVED** [review]
🔥

**Review by davran1001 — APPROVED** [review]

**Review by kuznetsov — APPROVED** [review]

**Pull request merged**